{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Yuvraj-Jaiswal/Vgg16-on-Chest-CT-scan/blob/main/Vgg16_Covid_CT.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "background_save": true,
          "base_uri": "https://localhost:8080/"
        },
        "id": "jjnCWvLTToYb",
        "outputId": "b202736f-06b9-42e9-d3ac-b275b99e4fb5"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Found 2281 images belonging to 2 classes.\n",
            "Found 200 images belonging to 2 classes.\n",
            "Epoch 1/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.4555 - accuracy: 0.7795\n",
            "Epoch 1: val_accuracy improved from -inf to 0.69000, saving model to Checkpoints\\model.01-0.69.h5\n",
            "143/143 [==============================] - 54s 352ms/step - loss: 0.4555 - accuracy: 0.7795 - val_loss: 0.5804 - val_accuracy: 0.6900\n",
            "Epoch 2/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.3379 - accuracy: 0.8536\n",
            "Epoch 2: val_accuracy did not improve from 0.69000\n",
            "143/143 [==============================] - 45s 303ms/step - loss: 0.3379 - accuracy: 0.8536 - val_loss: 0.6792 - val_accuracy: 0.6850\n",
            "Epoch 3/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.2829 - accuracy: 0.8829\n",
            "Epoch 3: val_accuracy improved from 0.69000 to 0.78000, saving model to Checkpoints\\model.03-0.78.h5\n",
            "143/143 [==============================] - 50s 321ms/step - loss: 0.2829 - accuracy: 0.8829 - val_loss: 0.4559 - val_accuracy: 0.7800\n",
            "Epoch 4/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.2614 - accuracy: 0.8873\n",
            "Epoch 4: val_accuracy did not improve from 0.78000\n",
            "143/143 [==============================] - 45s 302ms/step - loss: 0.2614 - accuracy: 0.8873 - val_loss: 0.4914 - val_accuracy: 0.7750\n",
            "Epoch 5/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.2491 - accuracy: 0.8922\n",
            "Epoch 5: val_accuracy did not improve from 0.78000\n",
            "143/143 [==============================] - 49s 314ms/step - loss: 0.2491 - accuracy: 0.8922 - val_loss: 0.6447 - val_accuracy: 0.7150\n",
            "Epoch 6/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.2389 - accuracy: 0.9036\n",
            "Epoch 6: val_accuracy did not improve from 0.78000\n",
            "143/143 [==============================] - 46s 304ms/step - loss: 0.2389 - accuracy: 0.9036 - val_loss: 0.5137 - val_accuracy: 0.7600\n",
            "Epoch 7/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.2203 - accuracy: 0.9110\n",
            "Epoch 7: val_accuracy did not improve from 0.78000\n",
            "143/143 [==============================] - 51s 332ms/step - loss: 0.2203 - accuracy: 0.9110 - val_loss: 0.5381 - val_accuracy: 0.7400\n",
            "Epoch 8/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1908 - accuracy: 0.9277\n",
            "Epoch 8: val_accuracy did not improve from 0.78000\n",
            "143/143 [==============================] - 45s 301ms/step - loss: 0.1908 - accuracy: 0.9277 - val_loss: 0.5824 - val_accuracy: 0.7750\n",
            "Epoch 9/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1810 - accuracy: 0.9246\n",
            "Epoch 9: val_accuracy improved from 0.78000 to 0.79500, saving model to Checkpoints\\model.09-0.80.h5\n",
            "143/143 [==============================] - 49s 324ms/step - loss: 0.1810 - accuracy: 0.9246 - val_loss: 0.5263 - val_accuracy: 0.7950\n",
            "Epoch 10/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1759 - accuracy: 0.9316\n",
            "Epoch 10: val_accuracy did not improve from 0.79500\n",
            "143/143 [==============================] - 45s 302ms/step - loss: 0.1759 - accuracy: 0.9316 - val_loss: 0.6478 - val_accuracy: 0.7750\n",
            "Epoch 11/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1739 - accuracy: 0.9290\n",
            "Epoch 11: val_accuracy did not improve from 0.79500\n",
            "143/143 [==============================] - 46s 307ms/step - loss: 0.1739 - accuracy: 0.9290 - val_loss: 0.6290 - val_accuracy: 0.7750\n",
            "Epoch 12/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1670 - accuracy: 0.9338\n",
            "Epoch 12: val_accuracy did not improve from 0.79500\n",
            "143/143 [==============================] - 46s 307ms/step - loss: 0.1670 - accuracy: 0.9338 - val_loss: 1.0719 - val_accuracy: 0.6500\n",
            "Epoch 13/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1706 - accuracy: 0.9277\n",
            "Epoch 13: val_accuracy did not improve from 0.79500\n",
            "143/143 [==============================] - 47s 303ms/step - loss: 0.1706 - accuracy: 0.9277 - val_loss: 0.4833 - val_accuracy: 0.7800\n",
            "Epoch 14/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1447 - accuracy: 0.9426\n",
            "Epoch 14: val_accuracy did not improve from 0.79500\n",
            "143/143 [==============================] - 52s 332ms/step - loss: 0.1447 - accuracy: 0.9426 - val_loss: 0.6396 - val_accuracy: 0.7800\n",
            "Epoch 15/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1398 - accuracy: 0.9434\n",
            "Epoch 15: val_accuracy did not improve from 0.79500\n",
            "143/143 [==============================] - 46s 308ms/step - loss: 0.1398 - accuracy: 0.9434 - val_loss: 0.6160 - val_accuracy: 0.7950\n",
            "Epoch 16/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1488 - accuracy: 0.9448\n",
            "Epoch 16: val_accuracy improved from 0.79500 to 0.81000, saving model to Checkpoints\\model.16-0.81.h5\n",
            "143/143 [==============================] - 51s 324ms/step - loss: 0.1488 - accuracy: 0.9448 - val_loss: 0.4762 - val_accuracy: 0.8100\n",
            "Epoch 17/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1529 - accuracy: 0.9456\n",
            "Epoch 17: val_accuracy did not improve from 0.81000\n",
            "143/143 [==============================] - 50s 334ms/step - loss: 0.1529 - accuracy: 0.9456 - val_loss: 0.6674 - val_accuracy: 0.7350\n",
            "Epoch 18/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1555 - accuracy: 0.9395\n",
            "Epoch 18: val_accuracy did not improve from 0.81000\n",
            "143/143 [==============================] - 46s 304ms/step - loss: 0.1555 - accuracy: 0.9395 - val_loss: 0.4806 - val_accuracy: 0.7950\n",
            "Epoch 19/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1303 - accuracy: 0.9452\n",
            "Epoch 19: val_accuracy did not improve from 0.81000\n",
            "143/143 [==============================] - 47s 302ms/step - loss: 0.1303 - accuracy: 0.9452 - val_loss: 0.5598 - val_accuracy: 0.7700\n",
            "Epoch 20/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1370 - accuracy: 0.9443\n",
            "Epoch 20: val_accuracy improved from 0.81000 to 0.82000, saving model to Checkpoints\\model.20-0.82.h5\n",
            "143/143 [==============================] - 50s 323ms/step - loss: 0.1370 - accuracy: 0.9443 - val_loss: 0.4312 - val_accuracy: 0.8200\n",
            "Epoch 21/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1235 - accuracy: 0.9487\n",
            "Epoch 21: val_accuracy did not improve from 0.82000\n",
            "143/143 [==============================] - 46s 303ms/step - loss: 0.1235 - accuracy: 0.9487 - val_loss: 0.4797 - val_accuracy: 0.8050\n",
            "Epoch 22/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1372 - accuracy: 0.9426\n",
            "Epoch 22: val_accuracy did not improve from 0.82000\n",
            "143/143 [==============================] - 48s 307ms/step - loss: 0.1372 - accuracy: 0.9426 - val_loss: 0.4102 - val_accuracy: 0.8150\n",
            "Epoch 23/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1048 - accuracy: 0.9601\n",
            "Epoch 23: val_accuracy did not improve from 0.82000\n",
            "143/143 [==============================] - 52s 332ms/step - loss: 0.1048 - accuracy: 0.9601 - val_loss: 0.6000 - val_accuracy: 0.7600\n",
            "Epoch 24/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1364 - accuracy: 0.9470\n",
            "Epoch 24: val_accuracy improved from 0.82000 to 0.84000, saving model to Checkpoints\\model.24-0.84.h5\n",
            "143/143 [==============================] - 50s 336ms/step - loss: 0.1364 - accuracy: 0.9470 - val_loss: 0.4175 - val_accuracy: 0.8400\n",
            "Epoch 25/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1306 - accuracy: 0.9505\n",
            "Epoch 25: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 46s 309ms/step - loss: 0.1306 - accuracy: 0.9505 - val_loss: 0.6460 - val_accuracy: 0.7700\n",
            "Epoch 26/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1507 - accuracy: 0.9391\n",
            "Epoch 26: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 50s 331ms/step - loss: 0.1507 - accuracy: 0.9391 - val_loss: 0.4239 - val_accuracy: 0.8150\n",
            "Epoch 27/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1021 - accuracy: 0.9579\n",
            "Epoch 27: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 50s 334ms/step - loss: 0.1021 - accuracy: 0.9579 - val_loss: 0.7962 - val_accuracy: 0.8000\n",
            "Epoch 28/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1087 - accuracy: 0.9570\n",
            "Epoch 28: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 46s 307ms/step - loss: 0.1087 - accuracy: 0.9570 - val_loss: 0.6587 - val_accuracy: 0.7600\n",
            "Epoch 29/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1256 - accuracy: 0.9505\n",
            "Epoch 29: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 43s 286ms/step - loss: 0.1256 - accuracy: 0.9505 - val_loss: 0.4726 - val_accuracy: 0.7950\n",
            "Epoch 30/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1032 - accuracy: 0.9540\n",
            "Epoch 30: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 45s 285ms/step - loss: 0.1032 - accuracy: 0.9540 - val_loss: 0.5638 - val_accuracy: 0.7500\n",
            "Epoch 31/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1161 - accuracy: 0.9509\n",
            "Epoch 31: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 43s 286ms/step - loss: 0.1161 - accuracy: 0.9509 - val_loss: 0.5590 - val_accuracy: 0.8050\n",
            "Epoch 32/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1000 - accuracy: 0.9601\n",
            "Epoch 32: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 43s 286ms/step - loss: 0.1000 - accuracy: 0.9601 - val_loss: 0.4312 - val_accuracy: 0.8150\n",
            "Epoch 33/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.1058 - accuracy: 0.9579\n",
            "Epoch 33: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 48s 320ms/step - loss: 0.1058 - accuracy: 0.9579 - val_loss: 0.5873 - val_accuracy: 0.7800\n",
            "Epoch 34/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.0910 - accuracy: 0.9632\n",
            "Epoch 34: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 43s 283ms/step - loss: 0.0910 - accuracy: 0.9632 - val_loss: 0.5823 - val_accuracy: 0.8050\n",
            "Epoch 35/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.0992 - accuracy: 0.9610\n",
            "Epoch 35: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 48s 310ms/step - loss: 0.0992 - accuracy: 0.9610 - val_loss: 0.6842 - val_accuracy: 0.7350\n",
            "Epoch 36/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.0963 - accuracy: 0.9632\n",
            "Epoch 36: val_accuracy did not improve from 0.84000\n",
            "143/143 [==============================] - 43s 282ms/step - loss: 0.0963 - accuracy: 0.9632 - val_loss: 0.6386 - val_accuracy: 0.8000\n",
            "Epoch 37/400\n",
            "143/143 [==============================] - ETA: 0s - loss: 0.0947 - accuracy: 0.9641"
          ]
        },
        {
          "ename": "KeyboardInterrupt",
          "evalue": "ignored",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-8-211855bd6904>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     72\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloss\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'binary_crossentropy'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0moptimizer\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'adam'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmetrics\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'accuracy'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 73\u001b[0;31m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtrain_data\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mepochs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m400\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mvalidation_data\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtest_data\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcallbacks\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mCheckpointSaver\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mTensorboard_loger\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mworkers\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m     \u001b[0mfiltered_tb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 64\u001b[0;31m       \u001b[0;32mreturn\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     65\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# pylint: disable=broad-except\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     66\u001b[0m       \u001b[0mfiltered_tb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_process_traceback_frames\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__traceback__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/keras/engine/training.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_batch_size, validation_freq, max_queue_size, workers, use_multiprocessing)\u001b[0m\n\u001b[1;32m   1429\u001b[0m               \u001b[0muse_multiprocessing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0muse_multiprocessing\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1430\u001b[0m               \u001b[0mreturn_dict\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1431\u001b[0;31m               _use_cached_eval_dataset=True)\n\u001b[0m\u001b[1;32m   1432\u001b[0m           \u001b[0mval_logs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m'val_'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mval\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mval_logs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1433\u001b[0m           \u001b[0mepoch_logs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval_logs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m     \u001b[0mfiltered_tb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 64\u001b[0;31m       \u001b[0;32mreturn\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     65\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# pylint: disable=broad-except\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     66\u001b[0m       \u001b[0mfiltered_tb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_process_traceback_frames\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__traceback__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/keras/engine/training.py\u001b[0m in \u001b[0;36mevaluate\u001b[0;34m(self, x, y, batch_size, verbose, sample_weight, steps, callbacks, max_queue_size, workers, use_multiprocessing, return_dict, **kwargs)\u001b[0m\n\u001b[1;32m   1714\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mtf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprofiler\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexperimental\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTrace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'test'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstep_num\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstep\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_r\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1715\u001b[0m               \u001b[0mcallbacks\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mon_test_batch_begin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstep\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1716\u001b[0;31m               \u001b[0mtmp_logs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtest_function\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterator\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1717\u001b[0m               \u001b[0;32mif\u001b[0m \u001b[0mdata_handler\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshould_sync\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1718\u001b[0m                 \u001b[0mcontext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masync_wait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/tensorflow/python/util/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    148\u001b[0m     \u001b[0mfiltered_tb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    149\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 150\u001b[0;31m       \u001b[0;32mreturn\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    151\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    152\u001b[0m       \u001b[0mfiltered_tb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_process_traceback_frames\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__traceback__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/def_function.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwds)\u001b[0m\n\u001b[1;32m    913\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    914\u001b[0m       \u001b[0;32mwith\u001b[0m \u001b[0mOptionalXlaContext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jit_compile\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 915\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    916\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    917\u001b[0m       \u001b[0mnew_tracing_count\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexperimental_get_tracing_count\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/def_function.py\u001b[0m in \u001b[0;36m_call\u001b[0;34m(self, *args, **kwds)\u001b[0m\n\u001b[1;32m    952\u001b[0m       \u001b[0;31m# In this case we have not created variables on the first call. So we can\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    953\u001b[0m       \u001b[0;31m# run the first trace but we should fail if variables are created.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 954\u001b[0;31m       \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_stateful_fn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    955\u001b[0m       \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_created_variables\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mALLOW_DYNAMIC_VARIABLE_CREATION\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    956\u001b[0m         raise ValueError(\"Creating variables on a non-first call to a function\"\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/function.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   2955\u001b[0m        filtered_flat_args) = self._maybe_define_function(args, kwargs)\n\u001b[1;32m   2956\u001b[0m     return graph_function._call_flat(\n\u001b[0;32m-> 2957\u001b[0;31m         filtered_flat_args, captured_inputs=graph_function.captured_inputs)  # pylint: disable=protected-access\n\u001b[0m\u001b[1;32m   2958\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2959\u001b[0m   \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/function.py\u001b[0m in \u001b[0;36m_call_flat\u001b[0;34m(self, args, captured_inputs, cancellation_manager)\u001b[0m\n\u001b[1;32m   1852\u001b[0m       \u001b[0;31m# No tape is watching; skip to running the function.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1853\u001b[0m       return self._build_call_outputs(self._inference_function.call(\n\u001b[0;32m-> 1854\u001b[0;31m           ctx, args, cancellation_manager=cancellation_manager))\n\u001b[0m\u001b[1;32m   1855\u001b[0m     forward_backward = self._select_forward_and_backward_functions(\n\u001b[1;32m   1856\u001b[0m         \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/function.py\u001b[0m in \u001b[0;36mcall\u001b[0;34m(self, ctx, args, cancellation_manager)\u001b[0m\n\u001b[1;32m    502\u001b[0m               \u001b[0minputs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    503\u001b[0m               \u001b[0mattrs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mattrs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 504\u001b[0;31m               ctx=ctx)\n\u001b[0m\u001b[1;32m    505\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    506\u001b[0m           outputs = execute.execute_with_cancellation(\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/execute.py\u001b[0m in \u001b[0;36mquick_execute\u001b[0;34m(op_name, num_outputs, inputs, attrs, ctx, name)\u001b[0m\n\u001b[1;32m     53\u001b[0m     \u001b[0mctx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mensure_initialized\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     54\u001b[0m     tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,\n\u001b[0;32m---> 55\u001b[0;31m                                         inputs, attrs, num_outputs)\n\u001b[0m\u001b[1;32m     56\u001b[0m   \u001b[0;32mexcept\u001b[0m \u001b[0mcore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_NotOkStatusException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     57\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ],
      "source": [
        "from tensorflow.keras.layers import Dense,Conv2D,MaxPooling2D,Dropout,Flatten\n",
        "from tensorflow.keras import Sequential\n",
        "from tensorflow.keras.preprocessing.image import ImageDataGenerator\n",
        "import numpy as np\n",
        "from tensorflow.python.keras.utils import data_utils\n",
        "from tensorflow.keras.callbacks import ModelCheckpoint,TensorBoard\n",
        "from tensorflow.keras.models import load_model\n",
        "from tensorflow.keras.applications import VGG16\n",
        "\n",
        "Shape = 224\n",
        "\n",
        "WEIGHTS_PATH_NO_TOP = ('https://storage.googleapis.com/tensorflow/'\n",
        "                       'keras-applications/vgg16/'\n",
        "                       'vgg16_weights_tf_dim_ordering_tf_kernels_notop.h5')\n",
        "\n",
        "Data_Generator = ImageDataGenerator(rotation_range=30,width_shift_range=0.1,\n",
        "                                    height_shift_range=0.1,shear_range=0.2,zoom_range=0.2,fill_mode='nearest'\n",
        "                                    ,horizontal_flip=False,rescale=1/255)\n",
        "\n",
        "train_data = Data_Generator.flow_from_directory(directory=\"/content/drive/MyDrive/Data/Data/Train\",\n",
        "                                             target_size=(Shape,Shape),batch_size=16,class_mode = 'binary')\n",
        "\n",
        "test_data = Data_Generator.flow_from_directory(directory=\"/content/drive/MyDrive/Data/Data/Test\",\n",
        "                                               target_size=(Shape,Shape),batch_size=16,class_mode = 'binary')\n",
        "\n",
        "CheckpointSaver = ModelCheckpoint(\"Checkpoints\\\\model.{epoch:02d}-{val_accuracy:.2f}.h5\",monitor='val_accuracy',mode='max',save_best_only=True,verbose=1)\n",
        "Tensorboard_loger = TensorBoard(log_dir='./logs')\n",
        "\n",
        "model = Sequential()\n",
        "\n",
        "model.add(Conv2D(filters=64,kernel_size=(3,3),input_shape=(Shape,Shape,3),activation='relu',padding='same'))\n",
        "model.add(Conv2D(filters=64,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(MaxPooling2D(pool_size=(2,2)))\n",
        "\n",
        "model.add(Conv2D(filters=128,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(Conv2D(filters=128,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(MaxPooling2D(pool_size=(2,2)))\n",
        "\n",
        "model.add(Conv2D(filters=256,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(Conv2D(filters=256,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(Conv2D(filters=256,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(MaxPooling2D(pool_size=(2,2)))\n",
        "\n",
        "model.add(Conv2D(filters=512,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(Conv2D(filters=512,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(Conv2D(filters=512,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(MaxPooling2D(pool_size=(2,2)))\n",
        "\n",
        "model.add(Conv2D(filters=512,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(Conv2D(filters=512,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(Conv2D(filters=512,kernel_size=(3,3),activation='relu',padding='same'))\n",
        "model.add(MaxPooling2D(pool_size=(2,2)))\n",
        "\n",
        "weights_path = data_utils.get_file(\n",
        "          'vgg16_weights_tf_dim_ordering_tf_kernels_notop.h5',\n",
        "          WEIGHTS_PATH_NO_TOP,\n",
        "          cache_subdir='models',\n",
        "          file_hash='6d6bbae143d832006294945121d1f1fc')\n",
        "\n",
        "model.load_weights(weights_path)\n",
        "\n",
        "for layer in model.layers:\n",
        "  layer.trainable = False\n",
        "\n",
        "model.add(Flatten())\n",
        "\n",
        "model.add(Dense(256,activation='relu'))\n",
        "model.add(Dense(64,activation='relu'))\n",
        "model.add(Dense(16,activation='relu'))\n",
        "model.add(Dense(1,activation='sigmoid'))\n",
        "\n",
        "model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])\n",
        "model.fit(train_data,epochs=400,validation_data=test_data,callbacks=[CheckpointSaver,Tensorboard_loger],workers=10)"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "name": "Vgg16-Covid-CT.ipynb",
      "provenance": [],
      "mount_file_id": "14otkKKDvAaHeeqL7H2U9l1ClXlDWsd4r",
      "authorship_tag": "ABX9TyP7XQ1CkrHUBWYwc9IGg14c",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}